### PR TITLE
Rotation based equalization

### PR DESCRIPTION
**Pull request by Giuseppe5**
## Reason for this PR
Implement rotation based equalization for weights and activation.

Highlights:
- Graph based
- Layerwise
- Merge Affine parameters of RMSNorm to make it compatible with rotation equalization
- Extend LLM entrypoint to support
- Tests

## Changes Made in this PR

The graph based region algorithm has been extended and generalized to support different set of supported layers based on what type of equalization is applied.

Important considerations:
- This PR depends on external library for fast implementation (I believe it should be made optional)
- This PR reuses code from other repositories (credited and with the appropriate license)

## Testing Summary

 Added one test that checks:
- Affine RMS norm parameter merging
- Rotation equalization being applied
- Mathematical invariance of the floating point output after rotation.

## Risk Highlight

- [x] This PR includes code from another work (please detail).
- [ ] This PR contains API-breaking changes.
- [ ] This PR depends on work in another PR (please provide links/details).
- [x] This PR introduces new dependencies (please detail).
- [ ] There are coverage gaps not covered by tests.
- [ ] Documentation updates required in subsequent PR.

## Checklist

- [x] Code comments added to any hard-to-understand areas, if applicable.
- [x] Changes generate no new warnings.
- [x] Updated any relevant tests, if applicable.
- [x] No conflicts with destination `dev` branch.
- [x] I reviewed my own code changes.
- [x] Initial CI/CD passing.
- [ ] 1+ reviews given, and any review issues addressed and approved.
- [ ] Post-review full CI/CD passing.

### CODE REVIEW

File: src/brevitas_examples/llm/main.py [hunk 1]
```diff
@@ -176,27 +177,41 @@ def main(args):
...
-    if args.eval:
-        assert args.export_target != 'torch_qcdq', "TorchScript QCDQ export and Evaluation simultaneously"
-        print("Float model eval...")
-        model = offload_model(model)
-        float_ppl = compute_perplexity(
-            model, validation_loader, context_length=args.seqlen // 2, tokenizer=tokenizer)
-        remove_hooks(model)
-        print(f"Float perplexity ({args.dataset}): {float_ppl:.3f}")
+    
+    # if args.eval:
```
Comment (Giuseppe5) [inline src/brevitas_examples/llm/main.py:181]: Restore

File: src/brevitas_examples/llm/main.py [hunk 2]
```diff
@@ -176,27 +177,41 @@ def main(args):
...
+            model, guards = torch._dynamo.export(model)(**calibration_loader[0])
         # Blockwise optimization does not work with FX at the moment
         args.gpxq_block_name = None
 
     # Apply LN affine merging before inserting MHA layers
     # since currently there is support only for merging into Linear
     if args.ln_affine_merge:
         print("Apply LN affine merge...")
-        apply_layernorm_affine_merge(model, dtype)
+        # apply_layernorm_affine_merge(model)
```
Comment (Giuseppe5) [inline src/brevitas_examples/llm/main.py:203]: Restore

File: tests/brevitas/graph/test_equalization.py [hunk 3]
```diff
@@ -18,210 +18,241 @@
...
-
-    if 'mha' in test_id:
-        in_shape = IN_SIZE_LINEAR
-    else:
-        in_shape = IN_SIZE_CONV
-
-    model_class = toy_model
-    model = model_class()
-    inp = torch.randn(in_shape)
+# def test_resnet18_equalization():
```
Comment (Giuseppe5) [inline tests/brevitas/graph/test_equalization.py:21]: Restore

File: src/brevitas/graph/equalize.py [hunk 4]
```diff
@@ -105,6 +123,12 @@ class WeightBiasWrapper:
     weight: torch.Tensor = None
     bias: torch.Tensor = None
 
+# Required for being hashable
+@dataclass(eq=True, frozen=True)
+class FunctionalNodeReference:
```
Comment (Giuseppe5) [inline src/brevitas/graph/equalize.py:128]: Still unused, maybe not needed?

**Comment by Giuseppe5** [comment]
Current limitations:
- No support for scaled_dot_product in a FX graph
- No support for MultiheadAttention
- Missing conversion from LayerNorm to RMSNorm to ensure rotation compatibility

File: src/brevitas_examples/llm/llm_quant/ln_affine_merge.py [hunk 5]
```diff
@@ -6,10 +6,23 @@
...
+from brevitas.graph.equalize import MergeLnAffine
 from brevitas.graph.utils import get_module
-from brevitas_examples.llm.llm_quant.run_utils import cast_to_float32
+
+
+def replace_rmsnorm_with_torch(model, config):
+    set_of_layers = set(type(x) for x in model.modules() if 'RMS' in type(x).__name__)
+    rewriters = [
+        ModuleToModuleByClass(
+            rms_cls, torch.nn.RMSNorm, normalized_shape=config.hidden_size, eps=config.rms_norm_eps)
```
Comment (nickfraser) [inline src/brevitas_examples/llm/llm_quant/ln_affine_merge.py:19]: torch version guard required?

**Review by nickfraser — CHANGES_REQUESTED** [review]
Could you add some basic tests for the LLM entry-point as well? Both for RMSNorm & rotations?

File: src/brevitas/graph/equalize.py [hunk 6]
```diff
@@ -1186,3 +1238,272 @@ def insert_mul_node(self, scale, shape, axis, act_node, batch_dim=0):
...
+    else:
+        return matmul_hadU(tensor)
+
+
+def _apply_ort_device(tensor, ort, *args):
+    ort = ort.type_as(tensor)
+    return torch.matmul(tensor, ort)
+
+
+def random_orthogonal_matrix(size):
```
[the remark this answers is not in the document]
Comment (Giuseppe5) [reply src/brevitas/graph/equalize.py:1257]: Yes

File: src/brevitas_examples/llm/llm_quant/ln_affine_merge.py (hunk 5 above)
Comment (Giuseppe5) [reply src/brevitas_examples/llm/llm_quant/ln_affine_merge.py:19]: Yes

**Comment by nickfraser** [comment]
Other than the hadamard matrix instantiation issue, this LGTM (once above changes are applied).

File: src/brevitas_examples/llm/main.py [hunk 7]
```diff
@@ -196,18 +199,34 @@ def main(args):
         remove_hooks(model)
         print(f"Float perplexity ({args.dataset}): {float_ppl:.3f}")
 
+    if args.replace_rmsnorm:
+        model = replace_rmsnorm_with_torch(model, model.config)
+
     if require_fx:
-        model = get_fx(model)
+        with torch.no_grad():
+            model, guards = torch._dynamo.export(model)(**calibration_loader[0])
```
Comment (Giuseppe5) [inline src/brevitas_examples/llm/main.py:207]: This has an impact on how dataloaders are created, we don't need extra kwargs for attention_mask

**Pull request merged**